### ISSUE TITLE
[MNG-2045] Maven can't compile against sibling test-jar dependency in multiproject (Test Attached)

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#4370]
**[Brian E Fox](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=brianf)** opened **[MNG-2045](https://issues.apache.org/jira/browse/MNG-2045?redirect=false)** and commented

I have 2 projects under a parent like so:

--Parent
--- sample-jar
--- sample-jar-user

sample-jar builds and installs a test-jar along with the normal jar. sample-jar-user depends on the test-jar at compile time. When I build from the parent folder, the build fails because it can't find the class. When I go to sample-jar-user and build, it works fine.

In the attached test case, to reproduce:
from the root folder, run mvn clean install - See it fail.
cd sample-jar-user; mvn clean install - see it succeed.

I remember reading somewhere that in multiprojects, maven attempts to locate the sibling classes in the source tree instead of using the jars from the repository. I'm guessing the problem is here that it's not looking in ../sample-jar/target/test-classes for this code, but really one should expect this to come from the repository.

---

**Affects:** 2.0.2

**Attachments:**
- [it1021.tar.gz](https://issues.apache.org/jira/secure/attachment/12711554/it1021.tar.gz) (_1.25 kB_)
- [mng-2045-ittest.zip](https://issues.apache.org/jira/secure/attachment/12712472/mng-2045-ittest.zip) (_10.28 kB_)
- [MNG-2045-maven-project-r577340.patch1](https://issues.apache.org/jira/secure/attachment/12712470/MNG-2045-maven-project-r577340.patch1) (_805 bytes_)
- [MNG-2045-maven-project-r577340.patch2](https://issues.apache.org/jira/secure/attachment/12712471/MNG-2045-maven-project-r577340.patch2) (_1.39 kB_)
- [release-log.zip](https://issues.apache.org/jira/secure/attachment/12713229/release-log.zip) (_150.38 kB_)
- [sample.zip](https://issues.apache.org/jira/secure/attachment/12711665/sample.zip) (_22.55 kB_)

**Issue Links:**
- [MSITE-432](https://issues.apache.org/jira/browse/MSITE-432) Incorrect classpath is used when site plugin/phase launches javadoc command
 (_**"is duplicated by"**_)
- [MRELEASE-140](https://issues.apache.org/jira/browse/MRELEASE-140) Tests fail during release:perform but work elsewhere

- [MJAR-68](https://issues.apache.org/jira/browse/MJAR-68) multi module release fails because of test-jar

- [MNG-3043](https://issues.apache.org/jira/browse/MNG-3043) Allow 'mvn test' to work with test-jar dependencies in a reactor

- [MNG-4032](https://issues.apache.org/jira/browse/MNG-4032) Test jar dependency not available for for main classes in multi module builds

- [MWAR-7](https://issues.apache.org/jira/browse/MWAR-7) Referenced projects' artifacts naming scheme behaves differently when ran from reactor

- [MNG-4056](https://issues.apache.org/jira/browse/MNG-4056) Relax resolution of active project artifacts from the reactor with regard to artifact type

- [MNG-2720](https://issues.apache.org/jira/browse/MNG-2720) Multiproject dependencies not accurate for project.compileClasspathElements when run from root project

- [MNG-2871](https://issues.apache.org/jira/browse/MNG-2871) Subartifact (ejb-client, test-jar etc.) are not reselved as active project artifacts in build phases prior to package

- [MNG-2877](https://issues.apache.org/jira/browse/MNG-2877) unable to resolve attached artifacts from reactor that are not in repo. (patch applied in svn and IT tests added)

- [MNG-3559](https://issues.apache.org/jira/browse/MNG-3559) Multi-Module Project: module that depends on sibling test jar cannot execute test-compile without install of sibling first

- [MNG-4248](https://issues.apache.org/jira/browse/MNG-4248) CLONE -Maven can't compile against sibling test-jar dependency in multiproject (Test Attached)
 (_**"is superceded by"**_)
- [MJAR-75](https://issues.apache.org/jira/browse/MJAR-75) [PATCH] Wrong artifact type attached to the project by the test-jar goal
 (_**"depends upon"**_)

11 votes, 13 watchers

**Comment by jira-importer** [comment]
**[Carlos Sanchez Gonzalez](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=carlos)** commented

Integration test.
If classes are moved to test folder and dependency set as test scope it works fine, using the jar from repo
If trying to use in compile scope it references ../child1/target/classes twice in the compiler classpath

**Comment by jira-importer** [comment]
**[Carlos Sanchez Gonzalez](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=carlos)** commented

Brian, it works if you use test-jar in a test scope, which is the usual case. Your main classes shuldn't depend on test classes from other modules

**Comment by jira-importer** [comment]
**[Brian E. Fox (imported)](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=brianf@infinity.nu)** commented

Yes, I realize this is an unusual case. The code is something we are working to fix, but it was working until I moved it into a common parent in preparation for some other stuff. It seems by extension this would be a problem for other dependencies that use classifiers.

**Comment by jira-importer** [comment]
**[Brian E. Fox (imported)](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=brianf@infinity.nu)** commented

I believe this is related to MWAR-7 and MOJO-286

**Comment by jira-importer** [comment]
**[Archimedes Trajano](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=trajano)** commented

We have the normal case, but it won't work unless we do an install first.

So mvn clean install on the project will run the tests correctly.
But mvn clean test won't because the test-jar dependency won't resolve.

**Comment by jira-importer** [comment]
**[Brian E. Fox (imported)](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=brianf@infinity.nu)** commented

This should be fixed. If you can reproduce, please make an IT test and attach. Instructions for an IT are here: http://docs.codehaus.org/display/MAVEN/Creating+a+Maven+Integration+Test

**Comment by jira-importer** [comment]
**[Brian E. Fox (imported)](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=brianf@infinity.nu)** commented

Patch and core it applied thanks to all that contributed. The IT test passes now.

**Comment by jira-importer** [comment]
**[Tomas Cirip](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ciript)** commented

Brian,

I find both sample projects (it1021.tar.gz and sample.zip) not working in 2.0.8). Can you comment on this, please? Thanks

Tomas

**Comment by jira-importer** [comment]
**[Brett Porter](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=brettporter)** commented

this was fixed in 2.0.8 - sample.zip, it1021.tar.gz and the integration all test with it.

Please open a new issue with more specifics if you are still seeing the problem.

**Comment by jira-importer** [comment]
**[Josh Chaitin-Pollak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jpollak)** commented

I don't think this bug is fixed.

If I download sample.zip and run 'mvn test-compile', I get the following:

jpollak@Challenger:~/src$ cd sample/
jpollak@Challenger:~/src/sample$ ls
pom.xml         sample-jar      sample-jar-user
jpollak@Challenger:~/src/sample$ mvn test-compile
[INFO] Scanning for projects...
[INFO] Reactor build order:
[INFO]   Unnamed - sample-project:sample-parent:pom:SNAPSHOT
[INFO]   Maven Quick Start Archetype
[INFO]   Maven Quick Start Archetype
[INFO] ------------------------------------------------------------------------
[INFO] Building Unnamed - sample-project:sample-parent:pom:SNAPSHOT
[INFO]    task-segment: [test-compile]
[INFO] ------------------------------------------------------------------------
[INFO] No goals needed for project - skipping
[INFO] ------------------------------------------------------------------------
[INFO] Building Maven Quick Start Archetype
[INFO]    task-segment: [test-compile]
[INFO] ------------------------------------------------------------------------
[INFO] [resources:resources]
[INFO] Using default encoding to copy filtered resources.
[INFO] [compiler:compile]
[INFO] Nothing to compile - all classes are up to date
[INFO] [resources:testResources]
[INFO] Using default encoding to copy filtered resources.
[INFO] [compiler:testCompile]
[INFO] Nothing to compile - all classes are up to date
[INFO] ------------------------------------------------------------------------
[INFO] Building Maven Quick Start Archetype
[INFO]    task-segment: [test-compile]
[INFO] ------------------------------------------------------------------------
[INFO] [resources:resources]
[INFO] Using default encoding to copy filtered resources.
[INFO] ------------------------------------------------------------------------
[ERROR] BUILD ERROR
[INFO] ------------------------------------------------------------------------
[INFO] Failed to resolve artifact.

Missing:

---

1) sample-project:sample-jar:test-jar:tests:SNAPSHOT

Try downloading the file manually from the project website.

Then, install it using the command:
mvn install:install-file -DgroupId=sample-project -DartifactId=sample-jar -Dversion=SNAPSHOT -Dclassifier=tests -Dpackaging=test-jar -Dfile=/path/to/file

Alternatively, if you host your own repository you can deploy the file there:
mvn deploy:deploy-file -DgroupId=sample-project -DartifactId=sample-jar -Dversion=SNAPSHOT -Dclassifier=tests -Dpackaging=test-jar -Dfile=/path/to/file -Durl=[url] -DrepositoryId=[id]

Path to dependency:
1) sample-project:sample-jar-user:jar:SNAPSHOT
2) sample-project:sample-jar:test-jar:tests:SNAPSHOT

---

1 required artifact is missing.

for artifact:
sample-project:sample-jar-user:jar:SNAPSHOT

from the specified remote repositories:
central (http://repo1.maven.org/maven2)

[INFO] ------------------------------------------------------------------------
[INFO] For more information, run Maven with the -e switch
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 1 second
[INFO] Finished at: Tue Apr 29 18:55:27 EDT 2008
[INFO] Final Memory: 3M/6M
[INFO] ------------------------------------------------------------------------

**Comment by jira-importer** [comment]
**[Ceki Gülcü](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ceki@qos.ch)** commented

I am seeing the same behavior with maven 2.0.9 when trying to build logback. See
http://svn.qos.ch/viewvc/logback/trunk/ if you would like to check out the project.

In this particular case, logback-classic depends on logback-core as follows:

    <dependency>
      <groupId>ch.qos.logback</groupId>
      <artifactId>logback-core</artifactId>
      <scope>compile</scope>
    </dependency>
    
    <dependency>
      <groupId>ch.qos.logback</groupId>
      <artifactId>logback-core</artifactId>
      <classifier>tests</classifier>
      <scope>test</scope>
    </dependency>

For a pristine installation, 'mvn package' fails with

---

Missing:

---

1) ch.qos.logback:logback-core:jar:tests:0.9.10

Try downloading the file manually from the project website.
[snip]

Path to dependency:
1) ch.qos.logback:logback-classic:jar:0.9.10
2) ch.qos.logback:logback-core:jar:tests:0.9.10

---

whereas 'mvn install' works fine.

The SLF4J project which I maintain suffers from the same problem. See http://svn.slf4j.org/repos/slf4j/trunk/ is you wish to check out SLF4J.

**Comment by jira-importer** [comment]
**[Joerg Schaible](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=joehni)** commented

For whatever it is worth, you may define the dependency as

```xml
<dependency>
    <groupId>ch.qos.logback</groupId>
    <artifactId>logback-core</artifactId>
    <type>test-jar</type>
    <scope>test</scope>
</dependency>
```

at least such a declaration works in quite some of our projects.

**Comment by jira-importer** [comment]
**[Ceki Gülcü](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ceki@qos.ch)** commented

Thank you Joerg. The change you suggested worked beautifully.

**Comment by jira-importer** [comment]
**[Benjamin Bentmann](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=bentmann)** commented

The documentation has been fixed in [r709688](http://svn.eu.apache.org/viewvc?view=rev&revision=709688). Left to discussion is whether the `<classifier>tests</classifier>` approach should also be supported.

**Comment by jira-importer** [comment]
**[Benjamin Bentmann](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=bentmann)** commented

The issue that was originally reported here, i.e. a bad compile class path when depending on a test JAR and running the install phase of a reactor build, was fixed in 2.0.8. Other issues like only running the test phase (MNG-3043) or the proper way to declare a test JAR dependency (MNG-4056) are addressed by the referenced JIRA tickets.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

I will have to correct my previous comment. I use the shade plugin. During perform it already uses the dependency reduced pom. All transitive dependencies for my test deps are removed. Watch out for: MSHADE-95
